### PR TITLE
chore(codecs): make source output types depend on codec

### LINKED ISSUE #12049: Source output types should depend on configured codec
Most of our sources currently hardcode the type of their output (e.g. log or metric). With the introduction of codecs, however, the correct value here is almost always dependent on the codec that they have configured. To make sure users do not encounter incorrect type checking errors, we should provide a simple way for codec configurations to identify the output type(s) they will produce, and make a pass through the sources to make sure that they are returning this value.

### PR DESCRIPTION
**Pull request by lukesteensen**
Closes #12049

This covers every source that has a `DeserializerConfig`, which I believe is the correct subset.

### CODE REVIEW

**Review by jszwedko — APPROVED** [review]
Sweet, this looks good to me! I wouldn't mind waiting for Pablo's review too though.

File: lib/codecs/src/decoding/format/bytes.rs [hunk 1]
```diff
@@ -25,6 +25,11 @@ impl BytesDeserializerConfig {
         BytesDeserializer::new()
     }
 
+    /// The data type of returned events
```
Comment (pablosichert) [inline lib/codecs/src/decoding/format/bytes.rs:28]: Just one tiny nitpit: These comments usually end with a `.` as they would end up as sentences in the HTML doc (if we configured that)

**Review by pablosichert — APPROVED** [review]
Sorry, been lagging behind with reviews a bit.

Nice, that solution is more straightforward than I would have thought!

File: lib/codecs/src/decoding/format/bytes.rs (hunk 1 above)
Comment (lukesteensen) [reply lib/codecs/src/decoding/format/bytes.rs:28]: Makes sense, I'll do that in a follow up.

**Pull request merged**